File: cpp/p21.cpp
#include "euler.h"

using namespace std;

int d( int n ) {
  vector<long long> proper_divisors = get_proper_divisors( n );	
  return accumulate( proper_divisors.begin(), proper_divisors.end(), 0 );
}

int main() {
  int sum = 0;

  for( int a = 1; a < 10000; a++ ) {
    int b = d( a );
		
    if( b > a && b < 10000 && d( b ) == a ) {
      sum += a;
      sum += b;
    }
  }
	
  cout << sum << endl;
}

File: clojure/p21.clj
(ns p21
  (:use [euler :only (proper-divisors sum sum-if timed-test)]))

(defn d [n]
  (sum (proper-divisors n)))

(defn amicable? [a]
  (let [b (d a)]
    (and (= (d b) a) (not= a b))))

(timed-test
 "Problem 21"
 31626
 (sum-if amicable? (range 220 10000)))
